### PR TITLE
chore: Plugin architecture improvements

### PR DESCRIPTION
**Pull request by kswenson**
In looking into implementation of the increase/decrease font size plugin for CODAP v3, I realized that there were limitations of the existing plugin API that made it more difficult than it needed to be, so I decided to do a refactor first. The root of the problem was that the toolbar button(s) required by a given plugin were configured by the toolbar code or by external clients, rather than by the plugin itself. This made it inconvenient to share state between the two pieces of code, for example for a click on the toolbar button to manipulate state owned by the plugin. I suspect this dates back to the original slate pre-0.50 plugin architecture, but there's no reason for the separation in the current slate environment and it makes a number of things harder.

Therefore, we now maintain a registry of toolbar button definitions and provide API for plugins to register their own toolbar button definitions. This allowed me to move a lot of plugin-specific code from the common toolbar code to the plugins themselves 😌, and also to eliminate the use of the event-emitter plugin (and its corresponding `eventemitter3` dependency) which was used to communicate between the two pieces of code.

Finally, a note on versions. In #83, I implied that the 0.11.0 release would require React 18. At that time, I released the 0.10.0 and 0.10.1 versions (which work on React 16+), but #83 was never merged and 0.11.0 was never publicly released. Given that CLUE is on React 17 and they might find these changes useful, I decided to switch gears and make 0.11.0 support React 17 and include these changes. Release 0.12.0 will be the release that requires React 18 and so #83 will be updated/merged after 0.11.0 has been released.

### CODE REVIEW

File: src/plugin-examples/variable-plugin.tsx [hunk 1]
```diff
@@ -119,16 +123,28 @@ function getDialogValuesFromNode(node?: CustomElement) {
 export function withVariables(editor: Editor) {
   registerVariableElement();
 
+  registerToolbarButtons(editor, [{
+    format: kVariableFormatCode,
+    SvgIcon: IconVariable,
+    tooltip: getPlatformTooltip("variable"),
+    isActive: () => !!editor?.isElementActive(kVariableFormatCode),
+    isEnabled: () => !!editor /* ?.isElementEnabled(kVariableFormatCode) */,
```
Comment (tealefristoe) [inline src/plugin-examples/variable-plugin.tsx:131]: Remove commented code?

File: src/plugins/list-plugin.tsx [hunk 2]
```diff
@@ -46,5 +50,23 @@ export function registerListBlocks() {
...
 export function withListBlocks(editor: Editor) {
   registerListBlocks();
+
+  registerToolbarButtons(editor, [
+    {
+      format: EFormat.numberedList,
+      SvgIcon: IconNumberedList,
+      tooltip: getPlatformTooltip("numbered list"),
+      isActive: () => !!editor && isBlockActive(editor, EFormat.numberedList),
+      onClick: () => toggleBlock(editor, EFormat.numberedList)
```
Comment (tealefristoe) [inline src/plugins/list-plugin.tsx:60]: I notice you're checking `!!editor` in similar places in `core-blocks-plugin.tsx` and `core-marks-plugin.tsx`. Can those be removed, or are similar checks necessary here?

**Review by tealefristoe — APPROVED** [review]
Looks good 👍

File: src/plugins/list-plugin.tsx (hunk 2 above)
Comment (kswenson) [reply src/plugins/list-plugin.tsx:60]: Good catch -- given that an `editor` is required to register the handler, I think we can remove the checks everywhere.

**Pull request merged**